### ISSUE TITLE
Picking up incorrect player in my team

### ISSUE THREAD

**Issue by ryanwilliams90** [alan-turing-institute/AIrsenal#277]
I am not sure what details are needed, but my problem is that for some reason the AIrsenal codebase does not pick up my team correctly. I swapped out a player last GW which does not seem to get reflected in my team and is not shown in my starting 11. It still shows the old player that has been subbed. I have removed the db from /tmp and started all the commands but have had no luck it does not pick up the change in player. 

Is there a way I can fix this to reflect my current team?

**Comment by jack89roberts** [comment]
Hi @ryanwilliams90 , have you run `airsenal_update_db`? If yes and it's still not working could you send your team ID so I can try to recreate the problem? Also, note that it won't pick up any transfers made in the current gameweek (this means you won't see any transfers made after 11:00 on 3rd October at the moment).

**Comment by ryanwilliams90** [comment]
Hey transfer after then I removed Zouma. My ID is: 354804. Hope am not missing something.

**Comment by jack89roberts** [comment]
I think that works fine for me but is it definitely the right ID? I don't see Zouma in the squad for any gameweek with that ID (in AIrsenal or on the website).

**Comment by ryanwilliams90** [comment]
It still put Zouma in my team?
== GK ==

Vicente Guaita (CRY)

== DEF ==

Andrew Robertson (LIV)
Enda Stevens (SHU)
George Baldock (SHU)

== MID ==

Mohamed Salah (LIV)(VC)
Fabio Henrique Tavares (LIV)
Lucas Rodrigues Moura da Silva (TOT)
Jarrod Bowen (WHU)

== FWD ==

Jordan Ayew (CRY)
Harry Kane (TOT)(C)
Jamie Vardy (LEI)

=== subs ===

Kurt Zouma (CHE)
Daniel Ceballos Fernández (ARS)
Seamus Coleman (EVE)
Ørjan Nyland (AVL)

**Comment by jack89roberts** [comment]
Could you check what value you get from this?
```bash
python -c "from airsenal.framework.utils import fetcher; print(fetcher.FPL_TEAM_ID)"
```
I'm expecting it not to be 354804, as the above looks like a completely different squad to [what I see on the web-site for that ID](https://fantasy.premierleague.com/entry/354804/event/4).

Have you tried to run it for multiple team IDs before, maybe? There are some issues with that at the moment.

**Comment by ryanwilliams90** [comment]
python -c "from airsenal.framework.utils import fetcher; print(fetcher.FPL_TEAM_ID)"
3548042

Sorry I have you the incorrect ID my apologies not sure where I copied that from. But My problem is still the same. Do we have a utils method for printing the team out?

I completely cleaned my env and have tried this on two different machines still the same problem.

**Comment by jack89roberts** [comment]
This should print what AIrsenal thinks your team is:
```
python -c "from airsenal.framework.optimization_utils import get_starting_squad; print(get_starting_team())"
```

I'll have another look at this later and see if I also have problems with that ID.

**Comment by ryanwilliams90** [comment]
I used get_starting_squad() but below is the print out.

python -c "from airsenal.framework.optimization_utils import get_starting_squad; print(get_starting_squad())"

=== starting 11 ===


== GK ==

Kepa Arrizabalaga (CHE)
Ørjan Nyland (AVL)

== DEF ==

Andrew Robertson (LIV)
Enda Stevens (SHU)
George Baldock (SHU)
Seamus Coleman (EVE)
Kurt Zouma (CHE)

== MID ==

Mohamed Salah (LIV)
Daniel Ceballos Fernández (ARS)
Fabio Henrique Tavares (LIV)
Lucas Rodrigues Moura da Silva (TOT)
Jarrod Bowen (WHU)

== FWD ==

Jordan Ayew (CRY)
Harry Kane (TOT)
Jamie Vardy (LEI)

=== subs ===

**Comment by nbarlowATI** [comment]
Hi @ryanwilliams90  I can reproduce this, doing `airsenal_setup_initial_db` and `airsenal_update_db` with your FPL_TEAM_ID .   Very strange indeed!!
Looking through the entries in the "transactions" table:
```
>>> from airsenal.framework.utils import *
>>> transactions = session.query(Transaction).all()
>>> for t in transactions:
>>>   print(t.gameweek, get_player(t.player_id).name, t.bought_or_sold)
```
it shows that John Egan was sold and Kurt Zouma was bought in GW 3, which looks like it's correct.   But then it shows that that John Egan was sold, and Romain Saiss was bought in GW 4, which is impossible as Egan shouldn't have been in the team.

Will investigate further...

**Comment by jack89roberts** [comment]
I noticed the calculated remaining budget for our (AIrsenal) squad seemed to be slightly wrong earlier - maybe it's that and it's causing removing/adding a player to fail?

**Comment by nbarlowATI** [comment]
So... it looks like the information from the FPL API is incorrect in this case!
If we do
```
>>> import requests
>>> r= requests.get('https://fantasy.premierleague.com/api/entry/3548042/transfers/')
>>> r.json()
```
we can see that in "event 4" (as in gameweek 4), there was "element_out": 353, and in "event: 3" there was also "element_out: 353")
and sure enough, 
```
python -c "from airsenal.framework.utils import *;print(get_player_from_api_id(353).name)"
```
confirms it's John Egan.

Not sure what we can do in this case, or if there's some way we can report to FPL that their API is giving bad information.
I guess you could add a Zouma -> Saiss substitution by hand:
```
>>> from airsenal.framework.utils import season, session
>>> from airsenal.framework.transaction_utils import *
>>> add_transaction(128, 4, -1, 50, CURRENT_SEASON, "AIrsenal"+CURRENT_SEASON, session)
>>> add_transaction(555, 4, 1, 50, CURRENT_SEASON, "AIrsenal"+CURRENT_SEASON, session)
>>> from airsenal.framework.optimization_utils import get_starting_squad
>>> get_starting_squad()
```
```

**Comment by jack89roberts** [comment]
Ahh I think this may actually be https://github.com/alan-turing-institute/AIrsenal/issues/200 rather than purely being an API bug - we currently fail to get the correct starting team after a free hit (@ryanwilliams90 used his free hit in gameweek 3, see here: https://fantasy.premierleague.com/api/entry/3548042/event/3/picks/)

I _think_ the gameweek 3 Egan -> Zouma was done as part of a free hit, then Egan is restored after the free hit, and finally Egan -> Saiss in GW4.

We probably should add a `is_free_hit` column or similar to the `Transaction` table to fix this? And sorry that past me was lazy and didn't fix this before!

**Comment by jack89roberts** [comment]
P.S. I think I hacked this before by deleting the free hit transactions from the database, so this should be an alternative way to fix things for you (but what @nbarlowATI suggests above will also work):
```
sqlite3 /tmp/data.db 'DELETE FROM "transaction" WHERE gameweek=3;'
```

**Comment by jack89roberts** [comment]
This should be fixed on the develop branch now (see #281 ). @ryanwilliams90 if you reinstall AIrsenal as follows (run from the parent AIrsenal directory) then delete the database and re-run everything I hope it should work as you expect:
```
git checkout develop
git pull
pip install --upgrade --force-reinstall .
```

**Comment by ryanwilliams90** [comment]
can confirm this fixed my issue thank you.
